### PR TITLE
Updates to docs, clarifying supported versions

### LINKED ISSUE #532: We should document plans for future releases
In particular, the [ongoing development](https://hypothesis.readthedocs.io/en/latest/development.html) page should include:

- Hypothesis' policy for which versions of Python are supported.  Also how and when we may decide to stop supporting older versions.

- A timeline for dropping Python 2 support - we keep mentioning 2020 but it's not documented anywhere.  [python3statement.org](http://www.python3statement.org) has a fantastic [practicalities page](http://www.python3statement.org/practicalities/) on how to make this work, especially if we want to make a py3-only release before the EOL date.

- ~~Hypothesis' policy for which versions of *Hypothesis* are supported.  Latest patch release only?  Latest patch of last *n* minor versions?  We don't need [diagrams](https://www.djangoproject.com/download/#supported-versions) or [extensive detail](https://docs.djangoproject.com/en/dev/internals/release-process/#supported-versions) like Django, but the basic approach is important information.~~

- Hypothesis' policy for deciding to release new major and minor versions of *Hypothesis*.  Feature accumulation?  Calendar schedule?  Deprecated features becoming a burden on development?  Probably related to #408.

### PR DESCRIPTION
**Pull request by Zac-HD**
Closes #532, replaces #603.

If it's not obvious what this changes, I should probably revise it 😜

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -79,6 +80,16 @@
 random.seed(0)
 '''
 
+# This config value must be a dictionary of external sites, mapping unique
+# short alias names to a base URL and a prefix.
+# See http://sphinx-doc.org/ext/extlinks.html
+extlinks = {
+    'gh-link': ('https://github.com/HypothesisWorks/hypothesis-python/%s', ''),
```
Comment (froztbyte) [inline docs/conf.py]: the smallest of nits: worth keeping the dict visually alphabetical by key?

I don't know if the dict will grow much, but it is useful to be able to run a visual bisecting search on sets of a certain size.

File: docs/development.rst [hunk 2]
```diff
@@ -14,17 +14,48 @@ I am the primary author of Hypothesis.
...
+Hypothesis releases follow `semantic versioning <http://semver.org/>`_.
+
+We maintain backwards-compatibility wherever possible, and use deprecation
+warnings to mark features that have been superseded by a newer alternative.
+If you want to detect this, the :py:meth:`strict setting <hypothesis.settings.strict>`
+upgrades all Hypothesis warnings to errors.
+
+We use continuous deployment to ensure that you can always use our newest and
+shiniest features - every change to the source tree is automatically built and
+published on PyPi as soon as it's merged onto master.  Of course, this only
```
Comment (alexwlchan) [inline docs/development.rst]: nit: the name is "PyPI", with a capital "i".

Also, I'm not a fan of phrases like "Of course" in documentation. What's obvious to us may not be obvious to the reader.

File: docs/supported.rst [hunk 3]
```diff
@@ -11,19 +11,13 @@ for the details.
...
-and bugs that affect them *might* get fixed.
-
-Hypothesis also supports PyPy (PyPy3 does not work because it only runs 3.2 compatible
-code, but if and when there's a 3.3 compatible version it will be supported), and
-should support 32-bit and narrow builds, though this is currently only tested on Windows.
-
-Hypothesis does not currently work on Jython (it requires sqlite), though could feasibly
-be made to do so. IronPython might work but hasn't been tested.
+Hypothesis also supports PyPy2, and will support PyPy3 when there is a stable
+release suporting python 3.4+.  Hypothesis does not currently work on Jython
```
Comment (alexwlchan) [inline docs/supported.rst]: /s/suporting/supporting

Capitalise "Python".

File: setup.py [hunk 4]
```diff
@@ -74,6 +74,7 @@ def local_file(name):
     zip_safe=False,
     extras_require=extras,
     install_requires=install_requires,
+    python_requires='==2.7,>=3.4',
```
Comment (alexwlchan) [inline setup.py]: I don’t think this works (that’s why I didn’t try to exclude 3.3 in #603). Quoting [PEP 440](https://www.python.org/dev/peps/pep-0440/#version-specifiers):

> The comma (`","`) is equivalent to a logical **and** operator: a candidate version must match all given version clauses in order to match the specifier as a whole.

There’s no way in this syntax to allow Python 2.7 or Python 3.4+.

Comment (alexwlchan) [inline setup.py]: (And this is why all the tests have failed.)

Comment (Zac-HD) [inline setup.py]: Whoops, fixed.  I guess this is an example of why we should be very careful with #454.

Comment (DRMacIver) [inline setup.py]: I think for the purposes of #454 changes to setup.py should *definitely* count as affecting code.

Comment (Zac-HD) [inline setup.py]: Certainly - I meant more that we should distinguish between source changes to make a release, and source changes to skip running tests.

**Comment by Zac-HD** [comment]
@DRMacIver, I'd like a review from you in particular on two issues: (1) are you happy with the tone of the new docs, including the bits about commercial support; and (2) are you OK with dropping Python 2.6 and 3.3?

**Comment by alexwlchan** [comment]
> are you OK with dropping Python 2.6 and 3.3

Python 2.6 support has been dead for a while.

Personally, I’d want to allow people to install on 3.3 for now (it’s still in support until September, and I don’t think there’s anything we’re dying to get in 3.4?). If we do drop 3.3, there's some 3.3-specific code in `setup.py` that we should rip out while we’re there.

**Comment by DRMacIver** [comment]
> I don’t think there’s anything we’re dying to get in 3.4?

The build doesn't actually pass on 3.3, although I think only due to one minor failure in pretty-printing. I don't feel like we especially need to block it, but it also is unlikely to be entirely bugfree if people use it.

File: docs/development.rst [hunk 5]
```diff
@@ -14,17 +14,48 @@ I am the primary author of Hypothesis.
 5. Planning the general roadmap of the project
 6. Doing sponsored development on tasks that are too large or in depth for other people to take on
 
-So all new features must either be sponsored or implemented by someone else. That being said, I take a fairly active
-role in shepherding pull requests and helping people write a new feature (see `the
-contributing guidelines <https://github.com/HypothesisWorks/hypothesis-python/blob/master/CONTRIBUTING.rst>`_ for
-details and `this pull request
-<https://github.com/HypothesisWorks/hypothesis-python/pull/154>`_ for an example of how the process goes). This isn't
+So all new features must either be sponsored or implemented by someone else.
+That being said, I take a fairly active role in shepherding pull requests and
```
Comment (DRMacIver) [inline docs/development.rst]: Maybe s/I/we/ these days?

File: docs/development.rst [hunk 6]
```diff
@@ -14,17 +14,48 @@ I am the primary author of Hypothesis.
...
+helping people write a new feature (see :gh-file:`CONTRIBUTING.rst` for
+details and :pull:`154` for an example of how the process goes). This isn't
 "patches welcome", it's "I will help you write a patch".
 
-All enhancement tickets on GitHub are tagged with either `help-wanted <https://github.com/HypothesisWorks/hypothesis-python/labels/help-wanted>`_
-if I think they're viable for someone else to pick up or `for-a-modest-fee <https://github.com/HypothesisWorks/hypothesis-python/labels/for-a-modest-fee>`_ if
+All enhancement tickets on GitHub are tagged with either :gh-link:`help-wanted <labels/help-wanted>`
+if I think they're viable for someone else to pick up or :gh-link:`for-a-modest-fee <labels/for-a-modest-fee>` if
```
Comment (DRMacIver) [inline docs/development.rst]: I actually think we should just drop this section. It's not really experienced much uptake on either of these, and I've not been good at tagging issues with them for a while.

File: docs/conf.py [hunk 7]
```diff
@@ -79,6 +80,16 @@
 random.seed(0)
 '''
 
+# This config value must be a dictionary of external sites, mapping unique
+# short alias names to a base URL and a prefix.
+# See http://sphinx-doc.org/ext/extlinks.html
```
Comment (DRMacIver) [inline docs/conf.py]: This is nice, thanks.

File: docs/development.rst [hunk 8]
```diff
@@ -14,17 +14,48 @@ I am the primary author of Hypothesis.
...
+passing our extensive test suite.
+
+
+Project Roadmap
+===============
+
+Hypothesis does not have a long-term release plan.  However some visibility
+into our plans for future :doc:`compatibility <supported>` may be useful:
+
+- We like compatibility, and maintain it as far as practical.  This generally
```
Comment (DRMacIver) [inline docs/development.rst]: I'm not sure we *like* compatibility. ;-)

File: docs/development.rst (hunk 5 above)
Comment (Zac-HD) [inline docs/development.rst]: Updating the docs to reflect that Hypothesis is a team project would be nice, and also a fair bit more work than just replacing pronouns.  I'll come back another time.

File: setup.py [hunk 9]
```diff
@@ -74,6 +74,7 @@ def local_file(name):
     zip_safe=False,
     extras_require=extras,
     install_requires=install_requires,
+    python_requires='>=2.7, !=3.0.*, !=3.1.*, !=3.2.*',
```
Comment (alexwlchan) [inline setup.py]: Did we get a ruling against supporting 3.3? (Sorry, it’s been a busy week.)

Comment (Zac-HD) [inline setup.py]: No ruling against, but neither is there any move to fix broken tests.

I think documenting support for 2.7 & 3.4+ and allowing installs on 3.3 is a good compromise.

**Comment by Zac-HD** [comment]
I have a couple of other things to change in the docs.

**Comment by Zac-HD** [comment]
Whoops, noticed a problem in the links dict.  Will also push some other changes to fix broken links and a few other things I noticed since.

**Comment by Zac-HD** [comment]
Pushed new commits to rebase, fix a typo in new extlink.  Also ran linkcheck and made a few other small edits like consistently recommending `pip install` over older methods.

**Comment by Zac-HD** [comment]
Closes #184 as wontfix - if you are in package install hell, please try using a recent version of pip.

**Comment by Zac-HD** [comment]
Generally I prefer not to merge my own pulls, but with two approvals I'm doing it this time so I can avoid a conflict with #620.

**Pull request merged**